### ISSUE TITLE
Discovery for use of fieldsets, legends and headings on VA.gov forms

### ISSUE THREAD

**Issue by rsmithadhoc** [department-of-veterans-affairs/vets-design-system-documentation#2142]
## Description

The forms team currently has the following problems with form elements:

- Web components are not flexible in allowing more than a string of text for ‘hint’ text to be read by screen readers. Some web components are not reading out the hint text for assistive tech users at all.
- Currently only the radio & checkbox group allows a heading in the legend.
- Guidance is unclear as to how engineers should implement legends, fieldsets, headers, component hint text, and form instructions.
- Guidance is unclear on designer best practices.
- There is also an issue with how header & descriptions are read out. The main issue is that an ID outside of the shadow DOM can not be targeted from within the shadow DOM.

This discovery work is to look into:

- What can we do to make this more modular (and does that work with the shadow DOM)? The goal is to give them the tools they need and limit DST's continued maintenance.
- If not, what are some options that will achieve their goals?
- Are there short-term fixes we can implement, with an eye towards a more ideal long-term solution?

## Details

https://docs.google.com/document/d/1J2qhbyMuuVpoxIRFe5FbC6itU-vCl_v2gjSU1wiVQnc/edit?usp=sharing

## Tasks
- [ ] Explore ways to give the forms team the options to implement their patterns (a more modular/flexible approach)
- [ ] Establish possible short and long term solutions to address their need

## Acceptance Criteria
- [ ] Potential solutions have been explored and outlined

**Comment by caw310** [comment]
Adding to sprint 37 as unplanned work due to forms team request.

**Comment by rsmithadhoc** [comment]
## Forms Discovery

### Current problems

- Heading is currently inside of the fieldset/legend. We provide a 
- Instructional text is not being read by screen reader
- The forms team wants to implement a pattern of one question per page where the heading is the form question.


### What the forms team would like to do

- Heading with `id="form-question"`, this heading would not go in the `<fieldset>` element
- Set focus to the heading for a better form experience
- A div (or a list element) with `id="form-instructions"`
- A fieldset with `aria-labeledby="form-question form-instructions"`
- This would allow screen reader users to tab to the form element and it will read the heading and instructions.

### Examples and Testing

To test out some possible options, I used the radio button markup from USWDS with a basic Stencil component, [forms-discovery.zip](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/files/12860995/forms-discovery.zip). This doesn't cover every scenario, but tries to address the core of the problem.

#### Non-Working Options

The options below are some things I tested, but these do not read `form-question`, `form-instructions` when tabbing to the form element.

- **Option A** - Slots - these cannot see through shadow DOM 

    Component usage:
    ```html
    <my-component>
        <h3 slot="heading" id="form-question">Heading</h3>
    </my-component>
    ```

    Component internals:
    ```html
    ...
    <slot name="heading"></slot>

    <div class="usa-legend usa-legend" id="form-description">Select one historical figure</div>
    
    <fieldset class="usa-fieldset" aria-labelledby="form-question form-description">
    ...
    ```
- **Option B** - Separate elements - these cannot see through shadow DOM
    Component usage:
    ```html
    <h3 id="form-question">Heading</h3>
    <div class="usa-legend usa-legend" id="form-description">Select one historical figure</div>

    <my-component></my-component>
    ```

    Component internals:
    ```html
    ...
    <fieldset class="usa-fieldset" aria-labelledby="form-question form-description">
    ...
    ```

### Working Options

The following options read `form-question`, `form-instructions` with a screen reader when tabbing to the form element.

- **Option C** - Workaround - multiple elements, pass the text for the screen reader into the component
    Component usage:
    ```html
    <h3>Example Heading</h3>
    <my-component heading="Example Heading"></my-component>
    ```

    Component internals:
    ```js
    ...
    @Prop() heading: string;
    ...
    ```

    ```html
    <div class="sr-only" id="form-question">{this.heading}</div>

    <div class="usa-legend usa-legend" id="form-description">Select one historical figure</div>

    <fieldset class="usa-fieldset" aria-labelledby="form-question form-description">   
    ```
- **Option D** - Shadow off, using slot

    Component usage:
    ```html
    ...
    <my-component><h3 id="form-question" slot="heading">Heading</h3></my-component>
    ...
    ```

    Component internals:
    ```jsx
    ...
    shadow: false,
    ...
    ```
    ```html
    <slot name="heading"></slot>

    <div class="usa-legend usa-legend" id="form-description">Select one historical figure</div>
    ```
- **Option E** - All elements inside the component
    
    This is similar to what we have now and what we tend to do, but our current component structure is different. In ours, the label can be made a heading to sit inside of the fieldset.
    ```html
    ...
    <h3 id="form-question">Heading</h3>

    <div class="usa-legend usa-legend" id="form-description">Select one historical figure</div>

    <fieldset class="usa-fieldset" aria-labelledby="form-question form-description">
        ...
    ```
- **Option F** - Primitive Components

    Directly use components like `<va-radio-option>`, but teams can maintain the surrounding markup.

    Component usage:
    ```html
    ...
    <h3 id="form-question">Heading Example</h3>
    <div class="usa-legend usa-legend" id="form-description">Select one historical figure</div>

    <fieldset class="usa-fieldset" aria-labelledby="form-question form-description">         
      <div class="usa-radio">
        <va-radio-option id="historical-truth" name="historical-figures" value="sojourner-truth" checked>Sojourner Truth</va-radio-option>
      </div>
      ...
    ```

    Component internals:
    ```tsx
    import { Component, Host, h, Prop } from '@stencil/core';

    @Component({
    tag: 'va-radio-option',
    styleUrl: 'va-radio-option.css',
    shadow: true,
    })
    export class VaRadioOption {
        @Prop() name: string;
        @Prop() id: string;
        @Prop() value: string;
        @Prop() checked: boolean;

        render() {
            return (
            <Host>
                <input class="usa-radio__input" id={this.id} type="radio" name={this.name} value={this.value} checked={this.checked} />
                <label class="usa-radio__label" htmlFor={this.id}>
                    <slot/>
                </label>
            </Host>
            );
        }
    }
    ```

### Thoughts and Next Steps

- Due to the shadow DOM, none of these options seem great.
- Options C and E seem feasible, but require the design system team to implement and maintain these solutions. 
    - Our current form elements do not all support headings and the one that does (radio button), places it in the fieldset/legend, which we don't want. 
    - If we change the structure of the form element components, we may need to update instances where those components are used.
- A solution like Option D is flexible for teams to use, but would require some work to remove the shadow DOM and test that styles/components behave as expected. This also runs the risk of there being conflicts in the future. This may not be a feasible option since we likely want to retain that encapsulation.
- Option F seems viable and flexible. It may require buy-in from DST that we don't enforce the patterns, we provide consistently styled building blocks for teams to use.

While this covers the DST's options, it may be more advisable for the forms team to simply implement this pattern without the use of DST components. It would allow them to be independent in developing their solution, but would be more work for them, and they would need to ensure they are keeping styles in line with the design system.

**Comment by laflannery** [comment]
@rsmithadhoc (and other interested parties) I wanted to follow up that our team did try to implement option F for the radio button but once our engineer dug into it we weren't able to because the `va-radio-option` component needs the `va-radio` component to handle the onChange events. [Here's the comment in our ticket](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/15268#issuecomment-1782369649) for reference. So we are going to build without any components for now. I wanted to make sure I followed up as I know this was still an outstanding question on how teams might move forward with this.

**Comment by rsmithadhoc** [comment]
> @rsmithadhoc (and other interested parties) I wanted to follow up that our team did try to implement option F for the radio button but once our engineer dug into it we weren't able to because the `va-radio-option` component needs the `va-radio` component to handle the onChange events. [Here's the comment in our ticket](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/15268#issuecomment-1782369649) for reference. So we are going to build without any components for now. I wanted to make sure I followed up as I know this was still an outstanding question on how teams might move forward with this.

Thanks! That will be helpful as we move forward with a solution.

**Comment by humancompanion-usds** [comment]
@rsmithadhoc - Did we open new issues to track this work? The problems are not resolved so I'm not sure why this is closed.

**Comment by rsmithadhoc** [comment]
I created these issues for follow-up work:
- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2269
- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2270
- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2271
- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2272
- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2273
